File: src/python/json_rpc.py
#!/usr/bin/python2

import json
import requests
import sys

api_key = "f13ce949-af39-4cb7-90bd-838767c86ac6";
target_url = "https://api.random.org/json-rpc/1/invoke"

def pack_json(method, num_vals, min_val, max_val, replacement, uid):
    params = {}
    params["apiKey"] = api_key
    params["n"] = num_vals
    params["min"] = min_val
    params["max"] = max_val
    params["replacement"] = replacement

    request_obj = {}
    request_obj["jsonrpc"] = "2.0"
    request_obj["method"] = method
    request_obj["params"] = params
    request_obj["id"] = uid
    return json.dumps(request_obj)

def send_json_request(json_dump):
    req = requests.get(target_url, data=json_dump)
    req_json = req.json()
    reqs_left = req_json['result']['requestsLeft']
    rand_vals = req_json['result']['random']['data']
    rpc_package = [reqs_left]
    rpc_package.extend([r for r in rand_vals])
    return str(rpc_package).replace(",", "").replace("[","").replace("]","")

def main():
    method = sys.argv[1]
    num_vals = int(sys.argv[2])
    min_val = int(sys.argv[3])
    max_val = int(sys.argv[4])
    replacement = sys.argv[5]
    uid = int(sys.argv[6])
    json_dump = pack_json(method, num_vals, min_val, max_val, replacement, uid)
    print(send_json_request(json_dump))

if __name__ == "__main__":
    main()

File: src/cpp/JsonRPC.cpp
/* DOMINION
 * David Mally, Richard Roberts
 * JsonRPC.cpp
 * Defines functions for making and parsing JSON requests
 * to the Random.org JSON RPC API.
 */
#include <iostream>
#include <string>
#include <cstdio>
#include <vector>
#include "JsonRPC.h"
#include "Defs.h"

json_rpc::JsonResponse::JsonResponse(int reqsLeft, std::vector<int> randVals) {
    m_reqsLeft = reqsLeft;
    m_randVals = randVals;
}

json_rpc::JsonResponse json_rpc::GetJsonResult(std::string method, int numVals, int minVal,
                          int maxVal, bool replacement) {
    std::string cmd = "python2 src/python/json_rpc.py ";
    cmd += method + " ";
    cmd += std::to_string(numVals) + " ";
    cmd += std::to_string(minVal) + " ";
    cmd += std::to_string(maxVal) + " ";
    cmd += std::to_string(replacement) + " ";
    cmd += std::to_string(baseId++);

    /* This section of code borrowed from StackOverflow */
    FILE *pipe = popen(cmd.c_str(), "r");
    if(!pipe) {
        std::vector<int> v;
        return json_rpc::JsonResponse(-1, v);
    }
    char buf[MAX_BUF_LEN];
    std::string result = "";
    while(!feof(pipe)) {
        if(fgets(buf, MAX_BUF_LEN, pipe) != NULL) {
            result += buf;
        }
    }
    pclose(pipe);
    /* This section of code borrowed from StackOverflow */

    int tmp = -1;
    int reqsLeft = -1;
    std::vector<int> randIndexes;

    // Parse number of API requests left
    std::string bufStr = result.substr(0, result.find(' '));
    sscanf(bufStr.c_str(), "%d", &reqsLeft);
    result = result.substr(result.find(' ') + 1, result.length());

    // Parse result of API call
    for(int i = 0; i < numVals; i++) {
        bufStr = result.substr(0, result.find(' '));
        sscanf(bufStr.c_str(), "%d", &tmp);
        if(result.find(' ') != std::string::npos) {
            result = result.substr(result.find(' ') + 1, result.length());
        } else {
            result = "";
        }
        randIndexes.push_back(tmp);
    }
    return json_rpc::JsonResponse(reqsLeft, randIndexes);
}

